File: examples/basics/read_state.cpp
// basics/read_state - read robot info, state, and motor status: GetRobotInfo /
// GetState / GetMotorStatus
//
// Usage: ./basics_read_state [robot_ip]

#include "sdk/robot.hpp"

#include <cstddef>
#include <cstdio>
#include <string>

int main(int argc, char **argv)
{
    const std::string robot_ip = (argc > 1) ? argv[1] : "";

    rcore::sdk::Robot robot;
    if (!robot.Initialize(robot_ip))
    {
        std::fprintf(stderr, "Initialize failed\n");
        return 1;
    }

    // Basic robot identity: model, serial number, and the SDK version.
    const auto info = robot.GetRobotInfo();
    std::printf("Robot model: %s  serial: %s  sdk: %s\n",
                info.robot_model.c_str(), info.robot_serial_number.c_str(),
                info.sdk_version.c_str());

    // Full state snapshot: joints, Cartesian pose, errors, and control mode.
    const auto state = robot.GetState();

    auto print_joints = [](const char *label, const auto &arr) {
        std::printf("%s:", label);
        for (size_t i = 0; i < arr.size(); ++i)
        {
            std::printf(" %.4f", arr[i]);
        }
        std::printf("\n");
    };

    // Joint-space state: the most commonly used fields.
    print_joints("Joint positions  [rad]  ", state.positions);
    print_joints("Joint velocities [rad/s]", state.velocities);
    print_joints("Joint torques    [N*m]  ", state.torques);
    print_joints("External torque  [N*m]  ", state.tau_ext);

    // Cartesian-space state (valid only when cartesian_valid is true).
    std::printf("Cartesian valid: %s\n", state.cartesian_valid ? "true" : "false");
    std::printf("TCP position    [m]:   %.4f %.4f %.4f\n",
                state.cartesian_position[0], state.cartesian_position[1],
                state.cartesian_position[2]);
    std::printf("TCP orientation [rad]: %.4f %.4f %.4f\n",
                state.cartesian_orientation[0], state.cartesian_orientation[1],
                state.cartesian_orientation[2]);

    // TCP velocity v=J*qd [vx,vy,vz,wx,wy,wz], expressed in the Base frame.
    print_joints("TCP velocity [m/s,rad/s]", state.cartesian_velocity);

    std::printf("Control mode: %d\n", static_cast<int>(state.control_mode));
    std::printf("Timestamp [s]: %.3f\n", state.timestamp);
    std::printf("Latched error count: %zu\n",
                state.errors.GetLatchedErrorCount());

    // Motor hardware status.
    const auto motor = robot.GetMotorStatus();
    std::printf("motor.enabled=%s estop=%s error=%s operational=%s\n",
                motor.enabled ? "true" : "false",
                motor.estop ? "true" : "false",
                motor.error ? "true" : "false",
                motor.operational ? "true" : "false");

    robot.Shutdown();
    return 0;
}

File: examples_py/basics/read_state.py
#!/usr/bin/env python3
"""basics/read_state - read robot info, state, and motor status.

Usage:
    python examples_py/basics/read_state.py [robot_ip]

Reads the most commonly used fields: robot identity, joint state, TCP pose and
velocity, control mode, and motor hardware status.
"""

import sys

from rcore_sdk import Robot


def fmt(values):
    return " ".join(f"{v:.4f}" for v in values)


def main():
    robot_ip = sys.argv[1] if len(sys.argv) > 1 else ""

    robot = Robot()
    if not robot.Initialize(robot_ip):
        print("Initialize failed", file=sys.stderr)
        return 1

    # Basic robot identity: model, serial number, and the SDK version.
    info = robot.GetRobotInfo()
    print(
        f"Robot model: {info['robot_model']}  "
        f"serial: {info['robot_serial_number']}  sdk: {info['sdk_version']}"
    )

    state = robot.GetState()

    # Joint-space state.
    print(f"Joint positions  [rad]  : {fmt(state.positions)}")
    print(f"Joint velocities [rad/s]: {fmt(state.velocities)}")
    print(f"Joint torques    [N*m]  : {fmt(state.torques)}")

    # Cartesian-space state (tvec / rvec are read-only properties).
    print(f"TCP position    [m]:   {fmt(state.tcp_pose.tvec)}")
    print(f"TCP orientation [rad]: {fmt(state.tcp_pose.rvec)}")

    # TCP velocity v=J*qd [vx,vy,vz,wx,wy,wz], expressed in the Base frame.
    print(f"TCP velocity [m/s,rad/s]: {fmt(state.cartesian_velocity)}")

    print(f"Control mode: {robot.GetControlMode()} (0 = Kinematics)")
    print(f"Timestamp [s]: {state.timestamp:.3f}")

    # Motor hardware status.
    motor = robot.GetMotorStatus()
    print(
        f"motor.enabled={motor.enabled} estop={motor.estop} "
        f"error={motor.error} operational={motor.operational}"
    )

    robot.Shutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
